File: apps/mobile/smarttaxi_app/test/map_presentation_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:maplibre_gl/maplibre_gl.dart' as native_map;
import 'package:smarttaxi_app/core/theme/app_theme.dart';
import 'package:smarttaxi_app/core/utils/map_layers.dart';

void main() {
  test('Native duplicate buildings use the supported visibility API', () async {
    for (final replacement in [
      'smarttaxi-3d-buildings',
      'smarttaxi-driver-3d-buildings',
    ]) {
      final controller = _LayerController(['building-3d', replacement]);
      await hideDuplicateLibertyBuildings(controller);
      expect(controller.visibilityChanges, [('building-3d', false)]);
    }
    final untouched = _LayerController(['building-3d']);
    await hideDuplicateLibertyBuildings(untouched);
    expect(untouched.visibilityChanges, isEmpty);
  });
  test('Liberty palette changes paint without changing geometry or visibility',
      () {
    for (final id in [
      'water',
      'building',
      'landuse_residential',
      'park',
      'landuse_school',
      'road_minor',
      'bridge_trunk_primary',
      'tunnel_motorway_casing',
      'waterway_river',
      'label_city',
      'poi_r1'
    ]) {
      final paint = libertyPresentationForLayer(id)!.toJson();
      expect(paint, isNotEmpty, reason: id);
      expect(
          paint.keys.every(
              (key) => key.endsWith('color') || key == 'text-halo-width'),
          isTrue,
          reason: id);
    }
    for (final id in [
      'smarttaxi-passenger-route-line',
      'road_one_way_arrow',
      'road_major_rail',
      'road_shield_us',
      'waterway_line_label',
      'custom-layer',
      'background'
    ]) {
      expect(libertyPresentationForLayer(id), isNull, reason: id);
    }
    expect(
        libertyPresentationForLayer('road_minor_casing')!
            .toJson()['line-color'],
        '#d9e2ef');
    expect(libertyPresentationForLayer('road_minor')!.toJson()['line-color'],
        '#ffffff');
  });

  test('Light and dark themes use the bundled typeface', () {
    for (final theme in [buildSmartTaxiTheme(), buildSmartTaxiDarkTheme()]) {
      expect(theme.textTheme.bodyMedium!.fontFamily, 'Inter');
      expect(theme.textTheme.titleLarge!.fontFamily, 'Inter');
    }
    expect(SmartTaxiColors.borderStrong, isNot(SmartTaxiColors.brand));
  });
}

class _LayerController extends Fake
    implements native_map.MapLibreMapController {
  _LayerController(this.ids);
  final List<String> ids;
  final visibilityChanges = <(String, bool)>[];

  @override
  Future<List<String>> getLayerIds() async => ids;

  @override
  Future<void> setLayerVisibility(String layerId, bool visible) async {
    visibilityChanges.add((layerId, visible));
  }
}

File: apps/web/tools/map-presentation.test.mjs
import assert from 'node:assert/strict';
import test from 'node:test';
import { libertyPaintForLayer, libertyPlaceLabelText, applyLibertyPresentation, hideDuplicateBuildings } from '../src/features/map/mapPresentation.mjs';

test('Liberty colors retain road geometry, widths, arrows and labels', () => {
  for (const id of ['background', 'water', 'building', 'landuse_residential', 'park', 'landuse_school', 'road_minor', 'bridge_trunk_primary', 'tunnel_motorway_casing', 'waterway_river', 'label_city', 'poi_r1']) {
    const paint = libertyPaintForLayer(id);
    assert(paint);
    assert(Object.keys(paint).every(key => key.endsWith('color') || key === 'text-halo-width'));
  }
  for (const id of ['smarttaxi-route', 'road_one_way_arrow', 'road_major_rail', 'road_shield_us', 'waterway_line_label', 'custom-layer']) assert.equal(libertyPaintForLayer(id), null);
  assert.equal(libertyPaintForLayer('road_minor_casing')['line-color'], '#d2deed');
  assert.equal(libertyPaintForLayer('road_minor')['line-color'], '#ffffff');
  assert.equal(libertyPaintForLayer('water')['fill-color'], '#b9d8f5');
  assert.equal(libertyPaintForLayer('label_city')['text-color'], '#50627a');
});

test('Custom styles are not recolored by a matching background id alone', () => {
  applyLibertyPresentation({ getStyle: () => ({layers: [{id: 'background'}]}), setPaintProperty: () => assert.fail('custom provider mutation'), setLayoutProperty: () => assert.fail('custom provider mutation') });
});

test('Liberty settlement labels use one local name with safe fallbacks', () => {
  assert.deepEqual(libertyPlaceLabelText, [
    'coalesce', ['get', 'name:nonlatin'], ['get', 'name:latin'], ['get', 'name_en'], ['get', 'name'],
  ]);
  const layers = [
    { id: 'landuse_residential', type: 'fill' },
    { id: 'road_one_way_arrow', type: 'symbol' },
    { id: 'building-3d', type: 'fill-extrusion' },
    { id: 'label_town', type: 'symbol', 'source-layer': 'place' },
    { id: 'poi_r1', type: 'symbol', 'source-layer': 'poi' },
  ];
  const layouts = [];
  applyLibertyPresentation({
    getStyle: () => ({ layers }),
    setPaintProperty() {},
    setLayoutProperty: (...args) => layouts.push(args),
  });
  assert.deepEqual(layouts, [['label_town', 'text-field', libertyPlaceLabelText]]);
});

test('Duplicate extrusion is hidden only after its replacement exists', () => {
  const layers = [
    {id: 'building-3d', type: 'fill-extrusion', source: 'tiles', 'source-layer': 'building'},
    {id: 'other-volume', type: 'fill-extrusion', source: 'tiles', 'source-layer': 'other'},
    {id: 'smarttaxi-3d-buildings', type: 'fill-extrusion', source: 'tiles', 'source-layer': 'building'},
  ];
  const changes = [];
  const map = { getStyle: () => ({layers}), getLayer: () => assert.fail('Runtime StyleLayer is not a style-spec object'), setLayoutProperty: (...args) => changes.push(args) };
  hideDuplicateBuildings(map, 'absent');
  assert.equal(changes.length, 0);
  hideDuplicateBuildings(map, 'smarttaxi-3d-buildings');
  assert.deepEqual(changes, [['building-3d', 'visibility', 'none']]);
});
